50645: Restoring Backups (#944)
replicatedhq/replicated-docs: docs/enterprise/snapshots-restoring-full.md

# Restoring from Backups

This topic describes how to restore full backups for disaster recovery on online and air gapped clusters. It also explains how to restore a partial backup (application only) from the Replicated admin console.

## Restore Any Backup from the CLI {#full-cli}

Using the kots CLI, you can restore any of the following backup types:

- **Full backups:** Includes the admin console and the application with its metadata for disaster recovery

> - **Partial backups:** Consists of the application and its metadata

Author: @sgalsaleh Is there a use case or scenario that you can recommend for partial backup and admin console only?

Reviewer: Here's what the current Creating Backups topic says about partial restores: "You can use partial backups for backing up information before deploying a new release version of an application, in case you need to roll back."

Nothing about admin console restores, however

> - **Admin console:** Consists of the admin console only

Author: Yes, indeed that is true. Good catch!

> Full backups are restored only with the CLI. During restoration, the admin console gets recreated and the admin console UI is disconnected during this process.

Reviewer: Ahh okay, I just made it down to the Admin Console section and I see that you are saying how you can _get the kots CLI commands from the admin console for full restores_, but you can't just click a button like you can for partial restores. 

So, you can restore from a _full backup_ from either the CLI or the admin console (since you can do a partial restore from a full backup). BUT you can only do a _full restore_ or _admin console restore_ with the CLI.

Partial backups can be restored using the CLI or the admin console. For more information about using the admin console to restore partial backups, see [Restore Partial Backups from the Admin Console](#admin-console).

### Existing Clusters

> If you are restoring to a healthy cluster, you can just run the `backup ls` and `restore` commands in the last two steps.

Reviewer: > If you are restoring to a healthy cluster, you can just run the `backup ls` and `restore` commands in the last two steps.

This sentence felt easy to miss, and since there are only three steps in the procedure anyway, maybe this can just be added to step 1.

Maybe beginning step 1 with "If you are not restoring to a healthy cluster"/ "If you are restoring to an unhealthy cluster". Then adding: "If you are restoring to a healthy cluster, skip this step and continue to step 2"

Alternatively, if you prefer to keep it outside the procedure, it could be more clear to say "...you can skip the first step of installing Velero in the cluster and continue to..."

Reviewer: (Same comment applies to the k8s installer procedures as well)

1. Install a version of Velero compatible with the one that was used to make the snapshot backup.
    * For restoring from an NFS or a host path storage destination, see [Configuring an NFS Storage Destination](snapshots-configuring-nfs) or [Configuring a Host Path Storage Destination](snapshots-configuring-hostpath) for the configuration steps and how to set up Velero.
    * For restoring from other storage destinations, see [Basic Install](https://velero.io/docs/v1.9/basic-install/) and [Plugins](https://velero.io/plugins/) in the Velero documentation.

    When you install, you must include these additional flags with the `velero install` command:

    - **Velero 1.10 and later:** Use the `--use-node-agent`, `--uploader-type=restic`, and `--use-volume-snapshots=false` flags.
    - **Velero versions earlier than 1.10:** Use the `--use-restic` and `--use-volume-snapshots=false` flags. 

1. Run the `kubectl kots backup ls` command to get a list of backups.

1. Run the following command to restore a full backup: 

    ```bash
    kubectl kots restore --from-backup BACKUP
    ```
    Replace `BACKUP` with the the name of the backup to restore from.
    

>     For more information about the available backup options, including partial backups and admin console only, see [restore](../reference/kots-cli-restore-index/) in _Reference_.

Reviewer: Need to remove the `..` at the start of this xref

1. Run the `kubectl kots backup ls` command to get a list of backups.

1. Run the following command to restore a full backup: 

    ```bash
    kubectl kots restore --from-backup BACKUP
    ```
    Replace `BACKUP` with the the name of the backup to restore from.
    
    For more information about the available backup options, including partial backups and admin console only, see [restore](../reference/kots-cli-restore-index/) in _Reference_.

### Air Gapped Kubernetes Installer Clusters

An air gapped Kubernetes installer-created cluster can be restored only if the storage backend used for backups is accessible from the new cluster.

The Kubernetes installer must also be able to assign the same IP address to the embedded private image registry in the new cluster.

To restore a backup on an air gapped Kubernetes installer cluster:

1. Run the following command to provide the Kubernetes installer with the correct registry IP address:

    ```bash
    cat install.sh | sudo bash -s airgap kurl-registry-ip=IP
    ```

    Replace `IP` with the registry IP address.

>     The registry from the old cluster does not need to be (and should not be) accessible.

Author: @sgalsaleh Can you clarify what is meant by "and should not be"? Does "should not" mean "must not" or "most likely is not"?

This language was carried over from a long time ago.

>     The registry from the old cluster does not need to be (and should not be) accessible.

Reviewer 2: I actually don't think this sentence is even needed. The registry is deployed and managed by the Kubernetes installer and there isn't really anything actionable by the user here other than providing the registry IP address.

    * **Other S3 Configuration (such as MinIO)**: See [velero configure-other-s3](../reference/kots-cli-velero-configure-other-s3/).
    * **NFS Configuration**: See [velero configure-nfs](../reference/kots-cli-velero-configure-nfs/) and [Configuring an NFS Storage Destination](snapshots-configuring-nfs).
    * **Host Path Configuration**: See [velero configure-hostpath](../reference/kots-cli-velero-configure-hostpath/) and [Configuring a Host Path Storage Destination](snapshots-configuring-hostpath).

1. Run the `kubectl kots backup ls` command to get a list of backups.

1. Run the following command to restore a full backup: 

    ```bash
    kubectl kots restore --from-backup BACKUP
    ```
    Replace `BACKUP` with the the name of the backup to restore from.
    
    For more information about the available backup options, including partial backups and admin console only, see [restore](../reference/kots-cli-restore-index/) in _Reference_.

## Restore Partial Backups from the Admin Console {#admin-console}

You can restore a partial backup from the admin console if your cluster is healthy. Partial backups consist of the application and its metadata.

>  The admin console deletes the selected application. All existing application manifests are removed from the cluster, and all `PersistentVolumeClaims` are deleted. This action is not reversible.

Author: Yep, moved it up.

  Then, all of the application manifests are redeployed to the namespace. All Pods are giving an extra `initContainer` and an extra directory named `.velero`, which are used to restore hooks. For more information about the restore process, see [Restore Reference](https://velero.io/docs/v1.9/restore-reference/) in the Velero documentation.

:::note
You can use the admin console to get the commands to restore a full backup or to restore only the admin console. Alternatively, you can use the kots CLI to restore any backup. See [Restore Any Backup from the CLI](#full-cli).
:::

To restore a partial backup from the admin console:

> 1. Select **Full Snapshots (Instance)** from the Snapshots tab.

Reviewer: Can you do a partial restore from a partial backup, instead of clicking "Full snapshots"?

Author: You can do a partial restore from a partial backup, but we've been wanting to steer them away from doing a partial backup. So I was steering them to the "Full snapshots" tab to get them used to the idea of being able to do a partial restore from a full backup.

I think i might be helpful if I clarify that making a full backup is the best practice and you can do a partial restore from the full backup. And I can clarify that if you are trying to restore from a partial backup, then click the "Partial snapshots" tab to do the partial restore from the partial backup.

# Restoring from Backups

This topic describes how to restore any type of backup from the kots CLI, and how to restore a partial backup (application only) from the Replicated admin console.

> Backups can be restored to healthy or unhealthy clusters, or brand new clusters.

Reviewer: I wonder if the phrase "brand new" might not translate directly to other languages/locales. 

If you decide to reword, you could also consider removing the phrase instead because it seems that the categories of "healthy or unhealthy" would include either new or existing clusters. (Though up to you, I'm not sure if it might be important to really highlight the new cluster piece)

> Backups can be restored to healthy or unhealthy clusters, or brand new clusters.

Reviewer 3: I meant to add that as I read the entirety of these docs, I realized that it's possible unhealthy is being used in a way that I just didn't understand. So knowing what "unhealthy" means in this situation would be good.

## About the Restore Process

> During any full or partial restore, the admin console deletes the selected application. All existing application manifests are removed from the cluster, and all `PersistentVolumeClaims` are deleted. This action is not reversible.

Author: No, a partial restore only restores an application, not the admin console. What's being described here is the restore process, in which the admin console (app manager really) deleted the application, etc. Then the restore process redeploys the application manifests.

This content was moved from here and reworded: https://docs.replicated.com/enterprise/snapshots-restoring-partial

If I've unintentionally garbled the meaning. please let me know @ajp-io

> During any full or partial restore, the admin console deletes the selected application. All existing application manifests are removed from the cluster, and all `PersistentVolumeClaims` are deleted. This action is not reversible.

Reviewer 3: I think you've reproduced it well. I just want to be sure that it's a true statement that _any_ full or partial restore deletes the selected application. Like if I'm only restoring the admin console, does it still do that? It might, I just don't know the specifics well enough.

This probably isn't worth spending too much time on. It was more of a musing I had.

> Then, all of the application manifests are redeployed. All Pods are given an extra `initContainer` and an extra directory named `.velero`, which are used to restore hooks. For more information about the restore process, see [Restore Reference](https://velero.io/docs/v1.9/restore-reference/) in the Velero documentation.

Author: I just added this as the first sentence in this section, which hopefully covers your most excellent point about admin console-only restores: "When you restore only the admin console, no changes are made to the application."

Also changed to "for restore hooks".

## Restore Any Backup from the CLI {#full-cli}

Using the kots CLI, you can do any of the following types of restores:

- **Full restore:** Restores the admin console and the application
- **Partial restore:** Restores the application

> - **Admin console:** Restores only the admin console

Reviewer: Very minor nit: the admin console bullet includes the word "only", but the partial restore bullet does not

> Full restores are only done with the CLI, although you can get the `restore` command from either the admin console or using the following procedures. 

Reviewer: >Full restores are only done with the CLI

Just noting some instances of passive voice throughout. If you feel like active would be more clear in any of the instances, could be worth a pass

Author: Done :)

> During a full restore or admin console restore, the admin console gets recreated and the admin console UI is disconnected during this process.

Reviewer: There are two "during" phrases in this sentence: "During a full restore or admin console restore," and then "during this process" at the end. I am not sure if these are referring to the same thing or not, but could consider removing one of the "during" phrases

Partial restores can be done using the CLI or the admin console. For more information about using the admin console, see [Restore the Application from the Admin Console](#admin-console). 

For CLI procedures, see the following sections:

- [Existing Clusters](#existing)
- [Online Kubernetes Installer Clusters](#online)
- [Air Gapped Kubernetes Installer Clusters](#air-gapped)

### Existing Clusters {#existing}

If you are restoring to a healthy cluster, you can skip reinstalling Velero and continue to running the `backup ls` and `restore` commands in the last two steps.

To restore a backup on an existing cluster:

> 1. (For new or unhealthy clusters) Install a version of Velero compatible with the one that was used to make the snapshot backup.

Reviewer 3: Same question about unhealthy here. I don't love that wording. Need to better understand what we're trying to convey with that.

> 1. (For new or unhealthy clusters) Install a version of Velero compatible with the one that was used to make the snapshot backup.

Reviewer: Minor nit: maybe a `:` here rather than a `.` to lead into the bulleted list?

>with the one that was used to make the snapshot backup:

    * For restoring from an NFS or a host path storage destination, see [Configuring an NFS Storage Destination](snapshots-configuring-nfs) or [Configuring a Host Path Storage Destination](snapshots-configuring-hostpath) for the configuration steps and how to set up Velero.

>     * For restoring from other storage destinations, see [Basic Install](https://velero.io/docs/v1.9/basic-install/) and [Plugins](https://velero.io/plugins/) in the Velero documentation.

Reviewer 3: Any reason we point to Velero's docs instead of here? https://deploy-preview-944--replicated-docs.netlify.app/enterprise/snapshots-storage-destinations

Author: It's a carry over, but with the recent improvements to the install, I think our docs are better/easier to follow now. I've changes the Xref. Thanks for questioning that.

    When you install, you must include these additional flags with the `velero install` command:

    - **Velero 1.10 and later:** Use the `--use-node-agent`, `--uploader-type=restic`, and `--use-volume-snapshots=false` flags.
    - **Velero versions earlier than 1.10:** Use the `--use-restic` and `--use-volume-snapshots=false` flags. 

> 1. Run the `kubectl kots backup ls` command to get a list of backups.

Reviewer 3: ```suggestion
1. Run the `kubectl kots get backups` command to get a list of backups.
```

1. Run the following command to restore a full backup: 

    ```bash
    kubectl kots restore --from-backup BACKUP
    ```
    Replace `BACKUP` with the the name of the backup to restore from.
    
    For more information about the available restore options, including application only and admin console only, see [restore](/reference/kots-cli-restore-index/) in _Reference_.

### Online Kubernetes Installer Clusters {#online}

If you are restoring to a healthy cluster, you can skip the installation and configuration steps and continue to running the `backup ls` and `restore` commands in the last two steps.

To restore a backup on a Kubernetes installer-created cluster:

> 1. (For new or unhealthy clusters) Provision a Kubernetes installer cluster and install the application. See [Installing with the Kubernetes Installer](installing-embedded-cluster).

Reviewer: >(For new or unhealthy clusters)

In other places I think we've done something like `(New or Unhealthy Clusters Only)` for these parenthetical phrases. Up to you though if you prefer "For" to "Only" in this case

1. (For new or unhealthy clusters) Run the following command to check whether Velero was installed as part of your installation script:

    ```bash
    velero version

>     ```

Reviewer: Suggestion: perhaps some context here on what to do if Velero was not installed? (I think the answer is that you just continue to the next step to Configure the storage destination and install at that point, which might be good to spell out)

> 1. (For new or unhealthy clusters) Configure a storage destination. See the following CLI documentation for your storage type:

Reviewer 3: Might want to mention that this needs to be the storage destination that holds the backup you want to use.

    * **AWS S3 Configuration**: See [velero configure-aws-s3](/reference/kots-cli-velero-configure-aws-s3/)
    * **Azure Configuration**: See [velero configure-azure](/reference/kots-cli-velero-configure-azure/)
    * **GCP Configuration**: See [velero configure-gcp](/reference/kots-cli-velero-configure-gcp/)
    * **Other S3 Configuration (such as MinIO)**: See [velero configure-other-s3](/reference/kots-cli-velero-configure-other-s3/)
    * **NFS Configuration**: See [velero configure-nfs](/reference/kots-cli-velero-configure-nfs/)

>     * **Host Path Configuration**: See [velero configure-hostpath](/reference/kots-cli-velero-configure-hostpath/)

Reviewer 3: Should we point to the individual CLI commands or the more detailed docs like https://deploy-preview-944--replicated-docs.netlify.app/enterprise/snapshots-storage-destinations and https://deploy-preview-944--replicated-docs.netlify.app/enterprise/snapshots-configuring-nfs, for example?

1. Run the `kubectl kots backup ls` command to get a list of backups.

1. Run the following command to restore a full backup: 

    ```bash
    kubectl kots restore --from-backup BACKUP
    ```
    Replace `BACKUP` with the the name of the backup to restore from.
    
    For more information about the available restore options, including application only and admin console only, see [restore](/reference/kots-cli-restore-index/) in _Reference_.

### Air Gapped Kubernetes Installer Clusters {#air-gapped}

To restore a backup on an air gapped Kubernetes installer cluster:

1. Run the following command to provide the Kubernetes installer with the correct registry IP address. The Kubernetes installer must be able to assign the same IP address to the embedded private image registry in the new cluster.

    ```bash

>     cat install.sh | sudo bash -s airgap kurl-registry-ip=IP

Reviewer 3: Based on the command, I assume this also installs a new cluster. That could be clearer.

    ```

    Replace `IP` with the registry IP address.

1. Use the kots CLI to configure Velero to use a storage destination. The storage backend used for backups must be accessible from the new cluster. 

    See the following CLI documentation for your storage type: 

    * **Other S3 Configuration (such as MinIO)**: See [velero configure-other-s3](/reference/kots-cli-velero-configure-other-s3/)
    * **NFS Configuration**: See [velero configure-nfs](/reference/kots-cli-velero-configure-nfs/)

>     * **Host Path Configuration**: See [velero configure-hostpath](/reference/kots-cli-velero-configure-hostpath/)

Reviewer 3: Same question about pointing to CLI docs or our docs on this subject.

1. Select **Full Snapshots (Instance)** from the Snapshots tab.

    ![Full Snapshot tab](/images/full-snapshot-tab.png)

    [View a larger image](/images/full-snapshot-tab.png)

1. Click the **Restore from this backup** icon (the circular blue arrows) for the backup that you want to restore.

1. In the **Restore from backup** dialog, select **Partial restore**.

    ![Restore Full Snapshot dialog](/images/restore-backup-dialog.png)

    :::note
    You can also get the CLI commands for full restores or admin console only restores from this dialog.
    :::     

1. At the bottom of the dialog, enter the application slug.

1. Click **Confirm and restore**.

### Use a Partial Backup to do a Partial Restore

To use a partial backup to do a partial restore in the admin console:

1. Select **Partial Snapshots (Application)** from the Snapshots tab.

    ![Partial Snapshot tab](/images/partial-snapshot-tab.png)

    [View a larger image](/images/partial-snapshot-tab.png)

1. Click the **Restore from this backup** icon (the circular blue arrows) for the backup that you want to restore.

    The **Restore from Partial backup (Application)** dialog opens.

> 1. Under **Type your application slug to continue**, enter the application slug next to the backup that you want to restore.

Reviewer 3: You've already selected which backup to use, so this doesn't seem right.
```suggestion
1. Under **Type your application slug to continue**, enter the application slug.
```

# Restoring from Backups

This topic describes how to restore any type of backup from the kots CLI, and how to restore a partial backup (application only) from the Replicated admin console.

Backups can be restored to healthy or unhealthy clusters, or brand new clusters.

## About the Restore Process

During any full or partial restore, the admin console deletes the selected application. All existing application manifests are removed from the cluster, and all `PersistentVolumeClaims` are deleted. This action is not reversible.

> Then, all of the application manifests are redeployed to the namespace. All Pods are giving an extra `initContainer` and an extra directory named `.velero`, which are used to restore hooks. For more information about the restore process, see [Restore Reference](https://velero.io/docs/v1.9/restore-reference/) in the Velero documentation.

Reviewer 2: ```suggestion
Then, all of the application manifests are redeployed. All Pods are given an extra `initContainer` and an extra directory named `.velero`, which are used to restore hooks. For more information about the restore process, see [Restore Reference](https://velero.io/docs/v1.9/restore-reference/) in the Velero documentation.
```

Reviewer 2: > All Pods are given an extra `initContainer` and an extra directory named `.velero`, which are used to restore hooks.

Is this sentence necessary? Seems like a Velero internal implementation detail.

1. Run the following command to restore a full backup: 

    ```bash
    kubectl kots restore --from-backup BACKUP
    ```
    Replace `BACKUP` with the the name of the backup to restore from.
    
    For more information about the available restore options, including application only and admin console only, see [restore](/reference/kots-cli-restore-index/) in _Reference_.

### Online Kubernetes Installer Clusters {#online}

If you are restoring to a healthy cluster, you can skip the installation and configuration steps and continue to running the `backup ls` and `restore` commands in the last two steps.

To restore a backup on a Kubernetes installer-created cluster:

1. (For new or unhealthy clusters) Provision a Kubernetes installer cluster and install the application. See [Installing with the Kubernetes Installer](installing-embedded-cluster).

1. (For new or unhealthy clusters) Run the following command to check whether Velero was installed as part of your installation script:

    ```bash
    velero version
    ```
1. (For new or unhealthy clusters) Do one of the following to configure a storage destination:

>     - If Velero is not installed, install Velero and configure an external storage destination. To start, see [Installing the Velero CLI](snapshots-velero-cli-installing).

Reviewer 2: ```suggestion
    - If Velero is not installed, install Velero and configure a storage destination. To start, see [Installing the Velero CLI](snapshots-velero-cli-installing).
```

Reviewer 2: So, for embedded clusters, the supported workflow is for the vendor to include the `velero` add-on in the Kubernetes installer. Otherwise, it might be the case that the customer's license doesn't support snapshots, so we shouldn't tell them to go install Velero by themselves. Maybe we can rewrite this part to only mention configuring a storage destination?

Author: Ah, ok, makes sense, as this is a restore. I was thinking about new installs, since the Velero add-on is something that the Vendor needs to add to the installer right now. So I thought that a custeomr might need to install it - but they might not have that included in their license. Anyway, I am mentioning only the "configure a storage destination" part now.

>     - If Velero is installed, configure an external storage destination. See the following CLI documentation for your storage type:

Reviewer 2: ```suggestion
    - If Velero is installed, configure a storage destination. See the following CLI documentation for your storage type:
```

        * **Host Path Configuration**: See [velero configure-hostpath](/reference/kots-cli-velero-configure-hostpath/)

1. Run the `kubectl kots backup ls` command to get a list of backups.

1. Run the following command to restore a full backup: 

    ```bash
    kubectl kots restore --from-backup BACKUP
    ```
    Replace `BACKUP` with the the name of the backup to restore from.
    
    For more information about the available restore options, including application only and admin console only, see [restore](/reference/kots-cli-restore-index/) in _Reference_.

### Air Gapped Kubernetes Installer Clusters {#air-gapped}

To restore a backup on an air gapped Kubernetes installer cluster:

1. Run the following command to provide the Kubernetes installer with the correct registry IP address. The Kubernetes installer must be able to assign the same IP address to the embedded private image registry in the new cluster.

    ```bash
    cat install.sh | sudo bash -s airgap kurl-registry-ip=IP
    ```

    Replace `IP` with the registry IP address.

    The registry from the old cluster does not need to be (and should not be) accessible.

> 1. Install the application in the cluster. See [Install in an Air Gapped Environment](installing-embedded-cluster#air-gap).

Reviewer 2: I don't think this step is necessary. Where did this come from?

Author: Ah, thanks. That was a gap in my understanding of this air gap process. I have removed the step.

    
    For more information about the available restore options, including application only and admin console only, see [restore](/reference/kots-cli-restore-index/) in _Reference_.

## Restore the Application from the Admin Console {#admin-console}

You can restore the application only, known as a _partial restore_, from either a full backup or a partial backup in the admin console.

Alternatively, you can use the kots CLI to restore any type of backup. See [Restore Any Backup from the CLI](#full-cli).

### Use a Full Backup to do a Partial Restore

To use a full backup to do a partial restore in the admin console:

1. Select **Full Snapshots (Instance)** from the Snapshots tab.

    ![Full Snapshot tab](/images/full-snapshot-tab.png)

    [View a larger image](/images/full-snapshot-tab.png)

1. Click the **Restore from this backup** icon (the circular blue arrows) for the backup that you want to restore.

1. In the **Restore from backup** dialog, select **Partial restore**.

    ![Restore Full Snapshot dialog](/images/restore-backup-dialog.png)

    :::note
    You can also get the CLI commands for full restores or admin console only restores from this dialog.
    :::     

> 1. In the list of available backups at the bottom of the dialog, enter the application slug in the **App Name** next to the backup that you want to restore. 

Reviewer 2: > enter the application slug in the **App Name** next to the backup that you want to restore. 

This sentence sounds a little bit off. I think "App Name" was the name of the application that you were testing with? Also, I don't believe there is a "list" of backups in the dialog, no?

Maybe:
```suggestion
1. At the bottom of the dialog, enter the application slug.
```
